### ISSUE TITLE
FakeItEasy.Core.FakeCreationException while creating (many) fakes

### ISSUE THREAD

**Issue by apacha** [FakeItEasy/FakeItEasy#773]
We are using FakeItEasy 2.0.0 and are creating thousands of fake objects during our unit-tests. Unfortunately we get a `FakeItEasy.Core.FakeCreationException` occasionally in different test-classes for different fake-objects on our CI:

```
MESSAGE:
SetUp : FakeItEasy.Core.FakeCreationException : 
  Failed to create fake of type "IConditionElementCollector".

  Below is a list of reasons for failure per attempted constructor:
    No constructor arguments failed:
      No usable default constructor was found on the type IConditionElementCollector.
      An exception of type System.InvalidOperationException was caught during this call. Its message was:
      Collection was modified; enumeration operation may not execute.
         bei System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
         bei System.Collections.Generic.List`1.Enumerator.MoveNextRare()
         bei System.Collections.Generic.List`1.Enumerator.MoveNext()
         bei Castle.DynamicProxy.Generators.Emitters.AbstractTypeEmitter.AddCustomAttributes(ProxyGenerationOptions proxyGenerationOptions)
         bei Castle.DynamicProxy.Generators.BaseProxyGenerator.CreateTypeAttributes(ClassEmitter emitter)
         bei Castle.DynamicProxy.Generators.ClassProxyGenerator.GenerateType(String name, Type[] interfaces, INamingScope namingScope)
         bei Castle.DynamicProxy.Generators.ClassProxyGenerator.<>c__DisplayClass1.<GenerateCode>b__0(String n, INamingScope s)
         bei Castle.DynamicProxy.Generators.BaseProxyGenerator.ObtainProxyType(CacheKey cacheKey, Func`3 factory)
         bei Castle.DynamicProxy.Generators.ClassProxyGenerator.GenerateCode(Type[] interfaces, ProxyGenerationOptions options)
         bei Castle.DynamicProxy.DefaultProxyBuilder.CreateClassProxyType(Type classToProxy, Type[] additionalInterfacesToProxy, ProxyGenerationOptions options)
         bei Castle.DynamicProxy.ProxyGenerator.CreateClassProxy(Type classToProxy, Type[] additionalInterfacesToProxy, ProxyGenerationOptions options, Object[] constructorArguments, IInterceptor[] interceptors)
         bei FakeItEasy.Creation.CastleDynamicProxy.CastleDynamicProxyGenerator.GenerateClassProxy(Type typeOfProxy, IEnumerable`1 argumentsForConstructor, IInterceptor interceptor, IEnumerable`1 allInterfacesToImplement)
         bei FakeItEasy.Creation.CastleDynamicProxy.CastleDynamicProxyGenerator.DoGenerateProxy(Type typeOfProxy, IEnumerable`1 additionalInterfacesToImplement, IEnumerable`1 argumentsForConstructor, IInterceptor interceptor)
         bei FakeItEasy.Creation.CastleDynamicProxy.CastleDynamicProxyGenerator.CreateProxyGeneratorResult(Type typeOfProxy, IEnumerable`1 additionalInterfacesToImplement, IEnumerable`1 argumentsForConstructor, IFakeCallProcessorProvider fakeCallProcessorProvider)


+++++++++++++++++++
STACK TRACE:
bei FakeItEasy.Core.DefaultExceptionThrower.ThrowFailedToGenerateProxyWithResolvedConstructors(Type typeOfFake, String reasonForFailureOfUnspecifiedConstructor, IEnumerable`1 resolvedConstructors)
bei FakeItEasy.Creation.FakeObjectCreator.TryCreateFakeWithDummyArgumentsForConstructor(Type typeOfFake, IProxyOptions proxyOptions, IDummyValueCreationSession session, String failReasonForDefaultConstructor, Boolean throwOnFailure)
bei FakeItEasy.Creation.FakeObjectCreator.CreateFake(Type typeOfFake, IProxyOptions proxyOptions, IDummyValueCreationSession session, Boolean throwOnFailure)
bei FakeItEasy.Creation.DefaultFakeAndDummyManager.CreateFake(Type typeOfFake, Action`1 optionsBuilder)
bei FakeItEasy.Creation.DefaultFakeCreatorFacade.CreateFake[T](Action`1 optionsBuilder)
bei FakeItEasy.A.Fake[T]()
bei VisibilityPropertyHandlerTest.Setup()
```

Any idea, how to fix this Heisenbug? The tests are running sequentially on the CI (no tests are running concurrently). It happens about once every week at different locations, usually where a lot of fakes are created rapidly: in the unit-tests.

**Comment by thomaslevesque** [comment]
Thanks for reporting this @apacha.

I think the problem is in `CastleDynamicProxyGenerator`. Apparently we update the list of `AdditionalAttributes` ([here](https://github.com/FakeItEasy/FakeItEasy/blob/2e8530324e98692b177834794104f467112f088a/src/FakeItEasy/Creation/CastleDynamicProxy/CastleDynamicProxyGenerator.cs#L45-L49)) on a shared instance of `ProxyGenerationOptions`, which causes [this loop](https://github.com/castleproject/Core/blob/eef7aaa4a76bccd4d33bd386ae317a33a035fcd8/src/Castle.Core/DynamicProxy/Generators/Emitters/AbstractTypeEmitter.cs#L102-L105) to fail if it happens at the same time. What is strange is that it really looks like a concurrency issue, but if you're running your tests sequentially, you should never hit it...

Anyway, if the root cause is what I think it is, it should be pretty easy to fix. The trouble is that it will be difficult to make sure it's fixed, since it can't be easily reproduced...

**Comment by apacha** [comment]
How about:

```
    class Program
    {
        static void Main(string[] args)
        {
            for (int i = 0; i < 50000; i++)
            {
                var fake = A.Fake<MyClass>();
            }
        }
    }

    internal interface MyClass
    {
    }
```

Can reproduce on every run ;-)

**Comment by apacha** [comment]
True. Even when creating lots of fakes concurrently, I can not reproduce this easily. I have no idea what's going on on our Jenkins server. Still, it would be a good thing to fix this, if possible.

**Comment by blairconrad** [comment]
@apacha, I'm a little confused. Your `Program` will fail because `MyClass` is internal. Unless you've letting your internals be visible to DynamicProxy somewhere else, and the example feels a little too light for that. I've tried here with a public `MyClass` and cannot see the failure.

**Comment by blairconrad** [comment]
@thomaslevesque, I think your analysis is spot on, but can't see any way for there to be a problem if tests are run serially. Only concurrent fake creation should be a problem, no?

**Comment by apacha** [comment]
My bad - the `Program `was garbage and I commented too quickly.

**Comment by thomaslevesque** [comment]
> Only concurrent fake creation should be a problem, no?

I think so... @apacha, your tests are run serially, but are you sure the tested code doesn't spawn threads that could indirectly create fakes (e.g. by calling an unconfigured method on a fake)?

**Comment by apacha** [comment]
That could definitely be the case. This particular test-case does no such things, but other tests in my project spawn tasks. 

@thomaslevesque it would be great if we could fix this, if it is that simple.

**Comment by blairconrad** [comment]
Writing a unit test to reproduce this may be nigh impossible, but we could create one that shows that we're mucking up our additional attributes by using static data. All we should need to do is make two Fakes on separate threads, with the fakes' constructors performing some waits, not unlike https://github.com/FakeItEasy/FakeItEasy/blob/492fab97d84d243415d6ab6f9daad7a7896fc267/tests/FakeItEasy.Tests/Core/ArgumentConstraintTrapTests.cs#L66.

**Comment by adamralph** [comment]
It sounds like we know what the problem is here and how to fix it. Marking as `ready`.

**Comment by thomaslevesque** [comment]
> It sounds like we know what the problem is here and how to fix it.

Well, we *think* we know... we haven't reproduced it yet.

**Comment by thomaslevesque** [comment]
> All we should need to do is make two Fakes on separate threads, with the fakes' constructors performing some waits, not unlike https://github.com/FakeItEasy/FakeItEasy/blob/492fab97d84d243415d6ab6f9daad7a7896fc267/tests/FakeItEasy.Tests/Core/ArgumentConstraintTrapTests.cs#L66.

I'm not sure it's that simple here... I think the issue happens in [this loop](https://github.com/castleproject/Core/blob/eef7aaa4a76bccd4d33bd386ae317a33a035fcd8/src/Castle.Core/DynamicProxy/Generators/Emitters/AbstractTypeEmitter.cs#L102-L105), and I don't see any way to pause its execution long enough to create another fake...

**Comment by blairconrad** [comment]
Drat, @thomaslevesque. I believe you're correct. The loop seems to be run before the fake's base constructor would be called. If the order had been different…

**Comment by thomaslevesque** [comment]
I managed to cause some concurrency related errors, but not the same as @apacha is seeing (it's an `IndexOutOfRangeException` on the call to `ProxyGenerationOptions.AdditionalAttributes.Clear()`). Also, in some runs I get many occurrences of this exception, and in others, none at all...

**Comment by blairconrad** [comment]
My experience exactly, @thomaslevesque.

**Comment by thomaslevesque** [comment]
I have a fix for the issue I mentioned above (I'll send a PR soon); hopefully, it will also fix @apacha's issue. If not, well, at least it fixes *something* ;)

However, since I haven't found a sure way to reproduce it, I don't have any test to prove that it works...

**Comment by adamralph** [comment]
Added to release notes.

**Comment by thomaslevesque** [comment]
Released in [2.2.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.2.0), thanks @apacha for reporting this!
